### PR TITLE
Fix handling native file watching thread exiting because of error

### PR DESCRIPTION
**Pull request by lptr**
When the native file watcher thread exits due to an error, we did stop waiting for more events coming in the queue, but did not stop relying on these events for detecting changes. Now we do.

Fixes https://github.com/gradle/gradle/issues/22467.

Bumps `native-platform` to https://github.com/gradle/native-platform/releases/tag/0.22-milestone-24 to include:
- https://github.com/gradle/native-platform/pull/320
- https://github.com/gradle/native-platform/pull/321

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherRegistryFactory.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
 import java.util.concurrent.BlockingQueue;
 import java.util.function.Predicate;
 
-public abstract class AbstractFileWatcherRegistryFactory<T extends AbstractFileEventFunctions<W>, W extends FileWatcher> implements FileWatcherRegistryFactory {
+public abstract class AbstractFileWatcherRegistryFactory<T extends AbstractNativeFileEventFunctions<W>, W extends FileWatcher> implements FileWatcherRegistryFactory {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherRegistryFactory.java:32]: Would it make sense to move `invalidateLogLevelCache()` to `AbstractFileEventFunctions`?

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherRegistryFactory.java:32]: Not really. It's a quirk of the native logging we have.

We could improve the logging to not be quirky, but that's not going to happen for 7.6.

[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherRegistryFactory.java:32]: It feels to me that we should move the `invalidateLogLevelCache()` to `AbstractFileEventFunctions` so we don't expose the native methods in the API. Though I don't feel very strongly about it.

**Review by wolfs — APPROVED** [review]
LGTM!

**Pull request merged**